src/main/java/org/commcare/core/network/ModernHttpRequester.java

```diff
@@ -41,149 +40,93 @@
-        HttpURLConnection httpConnection = (HttpURLConnection)builtUrl.openConnection();
-        setupConnectionInner(httpConnection);
-        if (isPostRequest) {
-            httpConnection.setRequestMethod("POST");
-            httpConnection.setDoOutput(true);
-            buildPostPayload(httpConnection);
-        } else {
-            httpConnection.setRequestMethod("GET");
+    /**
+     * Executes the HTTP Request. Can be called directly to bypass response processor.
+     *
+     * @return Response from the HTTP call
+     * @throws IOException
+     */
+    public Response<ResponseBody> makeRequest() throws IOException {
+        switch (method) {
+            case POST:
+                currentCall = commCareNetworkService.makePostRequest(url, params, addContentHeaders(requestBody), requestBody);
+                break;
+            case MULTIPART_POST:
+                currentCall = commCareNetworkService.makeMultipartPostRequest(url, params, headers, parts);
+                break;
+            case GET:
+                currentCall = commCareNetworkService.makeGetRequest(url, params, headers);
+                break;
+            default:
+                throw new IllegalArgumentException("Invalid HTTPMethod " + method.toString());
         }
-
-        return httpConnection;
-    }
-
-    private static void setupConnectionInner(HttpURLConnection httpConnection) {
-        httpConnection.setConnectTimeout(CONNECTION_TIMEOUT);
-        httpConnection.setReadTimeout(CONNECTION_SO_TIMEOUT);
-        httpConnection.setDoInput(true);
-        httpConnection.setInstanceFollowRedirects(true);
+        return currentCall.execute();
     }
 
-    private void buildPostPayload(HttpURLConnection httpConnection) throws IOException {
-        String paramsString = buildUrlWithParams().getQuery();
-        int bodySize = paramsString.length();
-        httpConnection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
-        httpConnection.setRequestProperty("Content-Length", bodySize + "");
-        // write to connection
-        OutputStream os = httpConnection.getOutputStream();
-        BufferedWriter writer = new BufferedWriter(
-                new OutputStreamWriter(os, "UTF-8"));
-        writer.write(paramsString);
-        writer.flush();
-        writer.close();
-        os.close();
-    }
 
-    protected URL buildUrlWithParams() throws MalformedURLException {
-        UriBuilder b = UriBuilder.fromUri(url.toString());
-        for (Map.Entry<String, String> param : params.entrySet()) {
-            b.queryParam(param.getKey(), param.getValue());
+    private Map<String, String> addContentHeaders(RequestBody postBody) {
+        headers.put("Content-Type", "application/x-www-form-urlencoded");
+        try {
+            headers.put("Content-Length", postBody.contentLength() + "");
+        } catch (IOException e) {
+            // Do we care ?
+            e.printStackTrace();
```

Comment:
I'd be somewhat surprised if this wasn't called in a context where adding the IOException to the method signature wasn't better than swallowing this exception.

Reply:
Again something I am retaining from earlier `ModernHTTPRequester` though I think it's due to the fact that not sending content-length as a header is not a deal breaker for a request to succeed.

Reply:
Totally fair to say that it was copied over from prior code, although no reason to not address error handling now that we're looking into it.

I think the point is less that the content length is mandatory for the request, but moreso that an IOException here doesn't indicate "no content length could be found", but indicates "There's a fundamental problem with some IO Read/Write that occurred while trying to find or set content length". 

If the pipe is broken or something and we swallow this I/O exception here, there's no guarantee that the next I/O exception thrown in the "real" mandatory context is going to have the same value of information that this one does.

If this was a Long Parse failure or something that is relevant _only_ to the content length I think swallowing sounds fine. Given that the error would be with the request objects themselves, i don't think this method can take responsibility for the failure.

Reply:
Makes sense. Different implementations of ContentBody can implement different length functions and so to be safe I am going to throw the exception from here.